### PR TITLE
Hive: Fix Deserializer to use source deserializer instead of the Iceberg ones

### PR DESCRIPTION
**Pull request by pvary**
During the testing of the write path @lcspinter run more extensive tests and identified an issue that fields which are supposed to use the WriteObjectInspector does not work.

The root cause was that the we both have to use the Hive generated ObjectInspectors to get the primitive object data from the Hive data model, and then use the IcebergObjectInspectors to convert them to the Iceberg data model.

The proposed solution creates an ObjectInspectorPair for every field which contains both the Hive and the Iceberg ObjectInspectors, and uses them when the conversion is needed.

Testing is done with the original code with changing the ObjectInspectors used. Also we hit #2053 so I had to disable the case sensitivity tests until #2053 is committed.

### CODE REVIEW

**Comment by pvary** [comment]
@marton-bod, @lcspinter: Could you please review?
Thanks,
Peter

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 1]
```diff
@@ -221,51 +223,54 @@ public ObjectInspector listElementPartner(ObjectInspector inspector) {
...
+      return sourceNameMap.get(originalName);
     }
+  }
 
-    @Override
-    public StructField getStructFieldRef(String fieldName) {
-      return nameMap.get(fieldName);
-    }
+  /**
+   * To get the data for Iceberg {@link Record}s we have to use the Hive ObjectInspectors (sourceInspector) to get
```
Comment (lcspinter) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:250]: nit: Could you please reformulate this doc? Just split up the long sentence into two parts :)

File: mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/WriteObjectInspector.java [hunk 2]
```diff
@@ -19,6 +19,11 @@
 
 package org.apache.iceberg.mr.hive.serde.objectinspector;
 
+/**
+ * Interface for converting the Hive primitive objects for to the objects which could be added to an Iceberg Record.
```
Comment (lcspinter) [inline mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/WriteObjectInspector.java:23]: typo:  ... for to ...

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 3]
```diff
@@ -78,62 +82,55 @@ Record deserialize(Object data) {
...
+    public FieldDeserializer primitive(PrimitiveType type, ObjectInspectorPair pair) {
+      return o -> {
+        if (o == null) {
+          return null;
+        }
+
+        ObjectInspector writerFieldInspector = pair.writerInspector();
+        ObjectInspector sourceFieldInspector = pair.sourceInspector();
+
+        Object result = ((PrimitiveObjectInspector) sourceFieldInspector).getPrimitiveJavaObject(o);
```
Comment (marton-bod) [inline mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:116]: Are we not losing here the `UUID.fromString` logic from before?
Edit: Ok now, I see it's been removed from the unit tests as well - can you give some more background on that? Thanks!

**Review by marton-bod — COMMENTED** [review]
Looks good, one question regarding uuid

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 3 above)
Comment (lcspinter) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:116]: I can answer this. My change (https://github.com/apache/iceberg/pull/2077) will introduce an IcebergUUIDObjectInspector to handle this conversion.

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java [hunk 4]
```diff
@@ -78,62 +82,55 @@ Record deserialize(Object data) {
...
+        if (writerFieldInspector instanceof WriteObjectInspector) {
```
[the remark this answers is not in the document]
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:117]: The original version was to use the Writable as an input for the `convert()`. In this new version `convert()` uses the native Hive object instead

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 3 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:116]: Yeah.. after #2077 we can have it back 😄

Comment (marton-bod) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java:116]: OK, thanks for clarifying!

File: mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java (hunk 1 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/Deserializer.java]: Tried my best 😄 Please check the new one.

File: mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/WriteObjectInspector.java (hunk 2 above)
Comment (pvary) [reply mr/src/main/java/org/apache/iceberg/mr/hive/serde/objectinspector/WriteObjectInspector.java]: Fixed. Thanks!

**Comment by lcspinter** [comment]
LGTM, +1 (non-binding). Thanks for the patch!

**Review by marton-bod — APPROVED** [review]

**Review by lcspinter — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @pvary! And thanks to @marton-bod and @lcspinter for the reviews.

**Pull request merged**